Source/DafnyCore/Verifier/Translator.cs

```diff
@@ -4343,48 +4385,56 @@ public enum StmtType { NONE, ASSERT, ASSUME };
       }
       builder.AddCaptureState(f.tok, false, "initial state");
 
-      DefineFrame(f.tok, f.Reads, builder, locals, null);
+      DefineFrame(f.tok, etran.ReadsFrame(f.tok), f.Reads, builder, locals, null);
       InitializeFuelConstant(f.tok, builder, etran);
 
+      // TODO: is this valid here given we haven't checked the well-formedness of the reads clauses yet?
```

Comment:
Please move this check after the well-formedness indeed so that the assertion won't implicitly assume well-formedness.

For example:
```
reads if a[0] then {} else {}
```

checking that the above is empty will automatically assume that a != null, which could be a soundness issue.

Reply:
I can't seem to create a test case that fails, and I suspect the ordering doesn't actually matter. But it does feel safer to reorder them so I'll still make the change.